java/algorithm/src/main/java/org/apache/arrow/algorithm/move/VectorDataCopier.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.move;
+
+import org.apache.arrow.vector.BaseFixedWidthVector;
+import org.apache.arrow.vector.BaseVariableWidthVector;
+import org.apache.arrow.vector.BitVectorHelper;
+
+import io.netty.buffer.ArrowBuf;
+import io.netty.util.internal.PlatformDependent;
+
+/**
+ * Utilities for moving data between vectors.
+ */
+public class VectorDataCopier {
+
+  /**
+   * Copy data for fixed-width vectors.
+   * @param srcVector the source vector.
+   * @param srcIndex the data index in the source vector.
+   * @param dstVector the destination vector.
+   * @param dstIndex the data index in the destination vector.
+   * @param typeWidth the width of the data type.
+   */
+  public static void fixedWidthDataCopy(
+          BaseFixedWidthVector srcVector, int srcIndex, BaseFixedWidthVector dstVector, int dstIndex, int typeWidth) {
+    if (srcVector.isNull(srcIndex)) {
+      BitVectorHelper.setValidityBit(dstVector.getValidityBuffer(), dstIndex, 0);
+    } else {
+      BitVectorHelper.setValidityBit(dstVector.getValidityBuffer(), dstIndex, 1);
+      PlatformDependent.copyMemory(srcVector.getDataBuffer().memoryAddress() + srcIndex * typeWidth,
+              dstVector.getDataBuffer().memoryAddress() + dstIndex * typeWidth, typeWidth);
+    }
+  }
+
+  /**
+   * Copy data for variable-width vectors.
+   * @param srcVector the source vector.
+   * @param srcIndex the data index in the source vector.
+   * @param dstVector the destination vector.
+   * @param dstIndex the data index in the destination vector.
+   */
+  public static void variableWidthDataCopy(
```

Comment:
it looks like this is unused except in the test?

Reply:
You are right. It is not currently used except in the test.
However, I think this should be a useful feature. So can we keep it?

Reply:
My concern with this is it seems like it corrupts all values after dstIndex?  If you only support linear appends it might pay to try to build something from a combination of a reader and a writer?

Reply:
This utility class is removed in the new commit. We use copyFrom method instead.